tracpro/contacts/models.py

```diff
@@ -127,53 +185,150 @@ def kwargs_from_temba(cls, org, temba_contact):
+    def save(self, *args, **kwargs):
+        self.name = self.name or ""  # RapidPro might return blank or null values.
+        return super(Contact, self).save(*args, **kwargs)
 
-    def push(self, change_type):
-        push_contact_change.delay(self.id, change_type)
 
-    def get_urn(self):
-        return tuple(self.urn.split(':', 1))
+class DataFieldQuerySet(models.QuerySet):
 
-    def get_responses(self, include_empty=True):
-        from tracpro.polls.models import Response
-        qs = self.responses.filter(pollrun__poll__is_active=True, is_active=True)
-        qs = qs.select_related('pollrun')
-        if not include_empty:
-            qs = qs.exclude(status=Response.STATUS_EMPTY)
-        return qs
+    def visible(self):
+        return self.filter(show_on_tracpro=True)
 
-    def release(self):
-        self.is_active = False
-        self.save()
-        self.push(ChangeType.deleted)
+    def by_org(self, org):
+        return self.filter(org=org)
+
+    def show_on_tracpro(self):
+        return self.update(show_on_tracpro=True)
+
+    def hide_on_tracpro(self):
+        return self.update(show_on_tracpro=False)
+
+
+class DataFieldManager(models.Manager.from_queryset(DataFieldQuerySet)):
+
+    def sync(self, org):
+        """Update the org's DataFields from RapidPro."""
+        # Retrieve current DataFields known to RapidPro.
+        temba_fields = {t.key: t for t in org.get_temba_client().get_fields()}
+
+        # Remove DataFields (and corresponding values per contact) that are no
+        # longer on RapidPro.
+        DataField.objects.by_org(org).exclude(key__in=temba_fields.keys()).delete()
+
+        # Create new or update existing DataFields to match RapidPro data.
+        for key, temba_field in temba_fields.items():
+            field, _ = DataField.objects.get_or_create(org=org, key=key)
+            field.label = temba_field.label
+            field.value_type = temba_field.value_type
+            field.save()
+
+        return temba_fields.keys()
+
+    def set_active_for_org(self, org, keys):
+        fields = DataField.objects.by_org(org)
+        fields.filter(key__in=keys).show_on_tracpro()
+        fields.exclude(key__in=keys).hide_on_tracpro()
+
+
+class DataField(models.Model):
+    """Custom contact data fields defined on RapidPro.
+
+    https://app.rapidpro.io/api/v1/fields
+    """
+    TYPE_TEXT = "T"
+    TYPE_DECIMAL = "N"
+    TYPE_DATETIME = "D"
+    TYPE_STATE = "S"
+    TYPE_DISTRICT = "I"
+    TYPE_CHOICES = (
+        (TYPE_TEXT, _("Text")),
+        (TYPE_DECIMAL, _("Decimal Number")),
+        (TYPE_DATETIME, _("Datetime")),
+        (TYPE_STATE, _("State")),
+        (TYPE_DISTRICT, _("District")),
+    )
+
+    org = models.ForeignKey("orgs.Org")
+    label = models.CharField(max_length=255, blank=True)
+    key = models.CharField(max_length=255)
+    value_type = models.CharField(max_length=1, choices=TYPE_CHOICES)
+    show_on_tracpro = models.BooleanField(default=False)
```

Comment:
I like that all the fields exist, with a show_on_tracpro boolean. Good idea, rather than only tracking those that the user selects. I wouldn't have thought to do that.